### ISSUE TITLE
Error message

### ISSUE THREAD

**Issue by thomasphansen** [ocelot-inc/ocelotgui#1]
Hello @pgulutzan ,

I'm trying (first time user! :) ) to debug a stored procedure, but I'm getting the following error:

```
mysql>$setup fix_username_duplicates;
Error 1064 (42000) You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '% PROCEDURE `test_th_roberto3`.`xxxmdbug002Pfix_username_duplicates`() NOT DETER' at line 1
```

Desktop info:
- Ocelotgui version: 1.0.0-1_i386 (deb package)
- Operating system: GNU/Linux (Linux Mint 17.3)
- Ocelotgui command line: ocelotgui --host=&lt;my internal IP for the server&gt; --user=&lh;my mysql user&gt;

(I am then clicking on File->Connect and providing the password)

Database server info:
- ODBS: MySQL 5.5.33
- Operating system: FreeBSD9.0 (amd64)

The stored procedure is as follows:

```
DELIMITER //

CREATE PROCEDURE 
    `fix_username_duplicates`()
BEGIN
    DECLARE p_bDoChanges BOOLEAN DEFAULT FALSE;
    DECLARE bDone BOOLEAN DEFAULT FALSE;
    DECLARE strUsername VARCHAR(50);
    DECLARE iCount INT;

    DECLARE cCursor CURSOR FOR
        SELECT
            count(temp_users.uUsername) as iCount,
            temp_users.uUsername
        FROM
            users AS temp_users
        WHERE
            temp_users.uUsername != ""
        GROUP BY
            temp_users.uUserName
        HAVING
            iCount > 1
    ;

    DECLARE CONTINUE HANDLER FOR NOT FOUND
        SET bDone = TRUE;

    CREATE TEMPORARY TABLE
        `temp_result_set`
        (
            `result_line` VARCHAR(250) NOT NULL DEFAULT ''
        );

    OPEN cCursor;

    rReadLoop: LOOP
    FETCH cCursor INTO iCount, strUsername;

        IF bDone THEN
            LEAVE rReadLoop;
        END IF;

        BEGIN
            DECLARE bDone2 BOOLEAN DEFAULT FALSE;
            DECLARE iUserID INT;
            DECLARE bInactive BOOLEAN;
            DECLARE strLastSeenDate DATETIME;

            DECLARE cCursor2 CURSOR FOR
                SELECT
                    users.uId               AS iUserID,
                    users.uis_deleted       AS bInactive,
                    users.uLastSeen         AS strLastSeenDate
                FROM
                    users
                WHERE
                    users.uUsername = strUsername
            ;

            OPEN cCursor2;

            rReadLoop2: LOOP
            FETCH cCursor2 INTO iUserID, bInactive, strLastSeenDate;
                IF bDone2 THEN
                    LEAVE rReadLoop;
                END IF;
                IF bInactive THEN
                    IF p_bDoChanges THEN
                        UPDATE
                            users
                        SET
                            users.uUsername = ""
                        WHERE
                            users.uId = iUserID;
                    ELSE
                        SELECT CONCAT("User #",iUserID,", username ", strUsername, ": inactive");
                        INSERT INTO
                            `temp_result_set`
                        SET
                            result_line = CONCAT("User #",iUserID,", username ", strUsername, ": inactive");
                    END IF;
                    LEAVE rReadLoop;
                END IF;
                IF strLastSeenDate = "0000-00-00 00:00:00" THEN
                    IF p_bDoChanges THEN
                        UPDATE
                            users
                        SET
                            users.uUsername = ""
                        WHERE
                            users.uId = iUserID;
                    ELSE
                        SELECT CONCAT("User #",iUserID,", username ", strUsername, ": never logged in");
                        INSERT INTO
                            `temp_result_set`
                        SET
                            result_line = CONCAT("User #",iUserID,", username ", strUsername, ": never logged in");
                    END IF;
                    LEAVE rReadLoop;
                END IF;
                IF strLastSeenDate < NOW() - INTERVAL 6 MONTH THEN
                    IF p_bDoChanges THEN
                        UPDATE
                            users
                        SET
                            users.uUsername = ""
                        WHERE
                            users.uId = iUserID;
                    ELSE
                        SELECT CONCAT("User #",iUserID,", username ", strUsername, ": no activities for the last 6 months");
                        INSERT INTO
                            `temp_result_set`
                        SET
                            result_line = CONCAT("User #",iUserID,", username ", strUsername, ": no activities for the last 6 months");
                    END IF;
                    LEAVE rReadLoop;
                END IF;

            END LOOP rReadLoop2;
            CLOSE cCursor2;

        END;
    END LOOP rReadLoop;
    CLOSE cCursor;
    SELECT
        *
    FROM
        `temp_result_set`;
    DROP TEMPORARY TABLE `temp_result_set`;
END//

DELIMITER ;
```

(p_bDoChanges was originally a parameter, but I couldn't figure out on your examples how to feed it ocelotgui, so I just converted it to a normal variable)

The $install; command runs fine (I can see the xxxmdbug database was created).

I am attaching the users table (used by the procedure).

Thanks for your help!
[users.sql.zip](https://github.com/ocelot-inc/ocelotgui/files/306369/users.sql.zip)

**Comment by thomasphansen** [comment]
UPDATE: I tried to reproduce your test case (as described at [this page](http://ocelot.ca/blog/blog/2015/03/02/the-ocelotgui-debugger/) ), and got almost the same error:

```
mysql>$setup fdemo, pdemo;
Error 1064 (42000) You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '% FUNCTION `test_th_roberto3`.`xxxmdbug003Ffdemo`() RETURNS int(11) NOT DETERMIN' at line 1
```

**Comment by pgulutzan** [comment]
I think your user+host name (what you see with "select current_user()") ends with "%". If you'll confirm that, then you have a temporary workaround (use a different login), and we'll fix it for the next release. But I see this bug in your stored procedure: you have a declared variable named "iCount" and you also have a column alias "as iCount", and you apparently hope that "HAVING iCount > 1" will check the column, but in fact it will look at the declared variable value, which is null. I'm sorry, the debugger wouldn't have helped you find that.

**Comment by thomasphansen** [comment]
Thank you, @pgulutzan, I've actually fixed the SP later on, using the good-old table logging debugging approach.
But regarding to ocelotgui's bug: The username is a pretty usual one (It's actually "spydev")! But I may have a clue for you: If I run

> show create procedure fix_username_duplicates;

, I get the declaration as follows:

>   CREATE DEFINER=`spydev`@`%` PROCEDURE `fix_username_duplicates`()
>       BEGIN
>          DECLARE p_bDoChanges BOOLEAN DEFAULT FALSE;
>   (...)

, which means mysql is storing the definition using the '%' wildcard. Maybe it's because I'm running on a remote mysql server, instead of a local one?
Anyway, if the "DEFINER" section is the problem, what if you remove it from the SP definition while parsing the procedure, before processing it? :)

**Comment by thomasphansen** [comment]
Hellp, @pgulutzan, I just found that explicitly adding DEFINER=spydev@localhost to the CREATE PROCEDURE statement allows me to successfully finish the $setup command! So, if you don't want to do anything regarding to the ' %'  wildcard, maybe you could add a note at ocelotgui documentation, telling that we should always add the DEFINER without a wildcard to prevent error, what do you think? :)
I've just run the debugger for the first time, and it looks terrific! ~~The only thing I really miss is a way of navigating through the history of commands from the command line (like we do with the up-down arrows on bash) - that would be a fantastic time-saver!~~ EDIT: Found the Ctrl+P and Ctrl+N actions! ;)

Thank you for the wonderful tool! I've made my day! :)

**Comment by pgulutzan** [comment]
What you say confirms that your user+host name is at some point for your particular setup seen by the server as ending in '%', so it's a bug (i.e. I should have anticipated this) rather than a documentation error, so we'll proceed to a fix in the next release, and then close this. The use of Ctrl+P and Ctrl+N is unfortunately not obvious, I'm glad you persevered to find them. Thank you for your kind words.

**Comment by pgulutzan** [comment]
Bug fix is in the source code, but won't appear in releases until the next release. In order for the bug fix to work, one must drop database xxxmdbug and all surrogate routines, then re-install.

**Comment by pgulutzan** [comment]
It is fixed in release 1.0.1.
